db/compatibility/compatibility_db_contents.go

```diff
@@ -0,0 +1,114 @@
+	// error -> if there is an error
+	GetCompatibilityData(ctx context.Context, tx db.Querier) (bool, T, error)
+	// SetCompatibilityData stores the compatibility data in the storage
+	// error -> if there is an error
+	SetCompatibilityData(ctx context.Context, tx db.Querier, data T) error
+}
+
+// RuntimeDataGetterFunc is a function that returns the runtime data
+type RuntimeDataGetterFunc[T comptypes.CompatibilityComparer[T]] func(ctx context.Context) (T, error)
+
+// CompatibilityCheck is the object that checks the compatibility between the runtime data and
+// the data stored in the storage. It's the implementation of the CompatibilityChecker interface
+type CompatibilityCheck[T comptypes.CompatibilityComparer[T]] struct {
+	RequireStorageContentCompatibility bool
+	RuntimeDataGetter                  RuntimeDataGetterFunc[T]
+	Storage                            CompatibilityDataStorager[T]
+	Logger                             common.Logger
+}
+
+// NewCompatibilityCheck creates a new CompatibilityCheck object
+func NewCompatibilityCheck[T comptypes.CompatibilityComparer[T]](
+	requireStorageContentCompatibility bool,
+	runtimeDataGetter RuntimeDataGetterFunc[T],
+	storage CompatibilityDataStorager[T]) *CompatibilityCheck[T] {
+	return &CompatibilityCheck[T]{
+		RequireStorageContentCompatibility: requireStorageContentCompatibility,
+		RuntimeDataGetter:                  runtimeDataGetter,
+		Storage:                            storage,
+	}
+}
+
+// Check checks the compatibility between the runtime data and the data stored in the storage
+func (s *CompatibilityCheck[T]) Check(ctx context.Context, tx db.Querier) error {
+	err := s.initialize()
+	if err != nil {
+		return fmt.Errorf("compatibilityCheck: fails to initialize. Err: %w", err)
+	}
+	runtimeData, err := s.RuntimeDataGetter(ctx)
+	if err != nil {
+		return err
+	}
+	// Read runtime data from storage
+	exists, storageData, err := s.Storage.GetCompatibilityData(ctx, tx)
+	// If there are no data in DB, we set the runtimeData
+	if !exists && err == nil {
+		// Store data
+		s.Logger.Infof("compatibilityCheck: no data stored, storing runtime data: [%s]", runtimeData.String())
+		return s.Storage.SetCompatibilityData(ctx, tx, runtimeData)
+	}
+	if err != nil {
+		return fmt.Errorf("compatibilityCheck: error reading value from storage. Err: %w", err)
+	}
+	// Compare data
+	err = runtimeData.IsCompatible(storageData)
+	if err != nil {
+		if s.RequireStorageContentCompatibility {
+			return fmt.Errorf("compatibilityCheck: data on DB is [%s] != runtime [%s]. Err: %w",
+				storageData.String(), runtimeData.String(), err)
+		} else {
+			s.Logger.Warnf("compatibilityCheck: data on DB is [%s] != runtime [%s]. Err: %w",
+				storageData.String(), runtimeData.String(), err)
+			return nil
+		}
+	}
+	s.Logger.Infof("compatibilityCheck: data in DB[%s] is compatible with [%s]",
+		storageData.String(), runtimeData.String())
+
+	return nil
+}
+
+func (s *CompatibilityCheck[T]) initialize() error {
```

Comment:
WDYT, in order to be semantically more correct?
```suggestion
func (s *CompatibilityCheck[T]) validate() error {
```

Reply:
In fact it's doing both things: the field Logger is initialized and the rest checked... I don't mind if you prefer `validate`  for me it's ok. WDYT?